File: src/parsers/linePrefix.ts
/**
 * Copyright 2025 IBM Corp.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Emitter } from "@/emitter/emitter.js";
import { entries } from "remeda";
import { ValueError } from "@/errors.js";
import { Serializable } from "@/internals/serializable.js";
import { shallowCopy } from "@/serializer/utils.js";
import { Cache } from "@/cache/decoratorCache.js";
import { ParserField } from "@/parsers/field.js";
import { Callback, InferCallbackValue } from "@/emitter/types.js";
import { ZodError } from "zod";
import { ValueOf } from "@/internals/types.js";
import { LinePrefixParserError } from "@/parsers/errors.js";
export * from "@/parsers/errors.js";

export interface ParserNode<T extends string, P extends ParserField<any, any>> {
  prefix: string;
  next: readonly T[];
  isStart?: boolean;
  isEnd?: boolean;
  field: P;
}

type Customizer<T extends NonNullable<unknown>, T2 extends NonNullable<unknown> = T> = (
  nodes: T,
  options: Options<T>,
) => { nodes: T2; options: Options<T2> };

type StringKey<T extends NonNullable<unknown>> = Extract<keyof T, string>;

interface Callbacks<T extends Record<string, ParserNode<StringKey<T>, ParserField<any, any>>>> {
  update: Callback<
    {
      [K in StringKey<T>]: {
        key: K;
        value: ParserField.inferValue<T[K]["field"]>;
        field: T[K]["field"];
      };
    }[StringKey<T>]
  >;
  partialUpdate: Callback<
    {
      [K in StringKey<T>]: {
        key: K;
        delta: string;
        value: ParserField.inferPartialValue<T[K]["field"]>;
        field: T[K]["field"];
      };
    }[StringKey<T>]
  >;
}

interface Line {
  value: string;
  newLine: boolean;
}

const NEW_LINE_CHARACTER = "\n" as const;

interface ExtractedLine<T extends NonNullable<unknown>> {
  key: StringKey<T>;
  value: string;
  partial: boolean;
}

const trimLeftSpaces = (value: string) => value.replace(/^\s*/g, "");
const linesToString = (lines: Line[]) =>
  lines.reduce(
    (acc, { newLine, value }) => `${acc}${newLine ? NEW_LINE_CHARACTER.concat(value) : value}`,
    "",
  );

interface Options<T extends NonNullable<unknown>> {
  fallback?: (value: string) => readonly { key: StringKey<T>; value: string }[];
  endOnRepeat?: boolean;
  waitForStartNode?: boolean;
  silentNodes?: StringKey<T>[];
}

type Input<K extends string = string> = Record<string, ParserNode<K, ParserField<any, any>>>;

export class LinePrefixParser<T extends Input<StringKey<T>>> extends Serializable {
  public readonly emitter = new Emitter<Callbacks<T>>({
    creator: this,
    namespace: ["agent", "parser", "line"],
  });

  protected readonly lines: Line[] = [];
  protected readonly excludedLines: Line[] = [];
  protected done = false;
  protected lastNodeKey: StringKey<T> | null = null;

  public readonly finalState = {} as LinePrefixParser.infer<T>;
  public readonly partialState = {} as Partial<Record<StringKey<T>, string>>;

  get isDone() {
    return this.done;
  }

  constructor(
    protected readonly nodes: T,
    protected readonly options: Options<T> = {},
  ) {
    super();

    let hasStartNode = false;
    let hasEndNode = false;

    for (const [key, { next, isStart, isEnd }] of entries(nodes ?? {})) {
      hasStartNode = Boolean(hasStartNode || isStart);
      hasEndNode = Boolean(hasEndNode || isEnd);

      for (const nextKey of next) {
        if (key === nextKey) {
          throw new ValueError(`Node '${key}' cannot point to itself.`);
        }
        if (!(nextKey in nodes)) {
          throw new ValueError(
            `Node '${key}' contains a transition to non-existing node '${nextKey}'.`,
          );
        }
      }
    }

    if (!hasStartNode) {
      throw new ValueError(`At least one start node must be provided!`);
    }
    if (!hasEndNode) {
      throw new ValueError(`At least one end node must be provided!`);
    }
  }

  fork<R extends T = T>(customizer: Customizer<T, R>) {
    const { nodes, options } = customizer(this.nodes, this.options);
    return new LinePrefixParser(nodes, options);
  }

  async add(chunk: string) {
    if (!chunk || this.done) {
      return;
    }

    chunk.split(NEW_LINE_CHARACTER).forEach((line, i, arr) => {
      const isFirstLine = i === 0;

      if (isFirstLine) {
        if (this.lines.length === 0) {
          this.lines.push({ newLine: false, value: line });
        } else {
          this.lines.at(-1)!.value += line;
        }
      } else {
        this.lines.push({ newLine: arr.length > 1, value: line });
      }
    });

    while (this.lines.length > 0) {
      const line = this.lines[0];
      const isLastLine = this.lines.length === 1;

      const lastNode = this.lastNodeKey ? this.nodes[this.lastNodeKey] : null;
      const isTerminationNode = lastNode
        ? Boolean(lastNode.isEnd && lastNode.next.length === 0)
        : false;

      const parsedLine =
        isTerminationNode || (lastNode && !line.newLine) ? null : this.extractLine(line.value);

      if (isLastLine && (parsedLine?.partial || !line.value)) {
        break;
      }
      this.lines.shift();

      if (parsedLine && !parsedLine.partial) {
        if (lastNode) {
          if (!lastNode.next.includes(parsedLine.key)) {
            if (parsedLine.key in this.finalState && this.options.endOnRepeat && lastNode.isEnd) {
              await this.end();
              return;
            }

            this.throwWithContext(
              `Transition from '${this.lastNodeKey}' to '${parsedLine.key}' does not exist!`,
              LinePrefixParserError.Reason.InvalidTransition,
              { line },
            );
          }

          await this.emitFinalUpdate(this.lastNodeKey!, lastNode.field);
        } else if (!this.nodes[parsedLine.key].isStart) {
          if (!this.options.waitForStartNode) {
            this.throwWithContext(
              `Parsed text line corresponds to a node "${parsedLine.key}" which is not a start node!`,
              LinePrefixParserError.Reason.NotStartNode,
              { line },
            );
          }
          this.excludedLines.push(line);
          continue;
        }

        const node = this.nodes[parsedLine.key];
        node.field.write(parsedLine.value);
        await this.emitPartialUpdate({
          key: parsedLine.key,
          value: node.field.getPartial(),
          delta: parsedLine.value,
          field: node.field,
        });
        this.lastNodeKey = parsedLine.key;
      } else if (this.lastNodeKey) {
        if (!this.nodes[this.lastNodeKey].field.raw) {
          line.value = trimLeftSpaces(line.value);
        }
        if (line.newLine) {
          line.value = `${NEW_LINE_CHARACTER}${line.value}`;
        }

        const node = this.nodes[this.lastNodeKey];
        node.field.write(line.value);
        await this.emitPartialUpdate({
          key: this.lastNodeKey,
          value: node.field.getPartial(),
          delta: line.value,
          field: node.field,
        });
      } else {
        this.excludedLines.push(line);
      }
    }
  }

  protected throwWithContext(
    message: string,
    reason: ValueOf<typeof LinePrefixParserError.Reason>,
    extra: { line?: Line; errors?: Error[] } = {},
  ): never {
    throw new LinePrefixParserError(
      [`The generated output does not adhere to the schema.`, message].join(NEW_LINE_CHARACTER),
      extra.errors ?? [],
      {
        reason,
        context: {
          lines: linesToString(this.lines.concat(extra.line ? [extra.line] : [])),
          excludedLines: linesToString(this.excludedLines),
          finalState: this.finalState,
          partialState: this.partialState,
        },
      },
    );
  }

  async end() {
    if (this.done) {
      return this.finalState;
    }

    if (!this.lastNodeKey && this.options.fallback) {
      const stash = linesToString([
        ...this.excludedLines.splice(0, Infinity),
        ...this.lines.splice(0, Infinity),
      ]);

      const nodes = this.options.fallback(stash);
      await this.add(
        nodes.map((node) => `${this.nodes[node.key].prefix}${node.value}`).join(NEW_LINE_CHARACTER),
      );
    }

    this.done = true;

    if (!this.lastNodeKey) {
      this.throwWithContext(
        "Nothing valid has been parsed yet!",
        LinePrefixParserError.Reason.NoDataReceived,
      );
    }

    const stash = linesToString(this.lines);
    this.lines.length = 0;

    const field = this.nodes[this.lastNodeKey].field!;
    if (stash) {
      field.write(stash);
      await this.emitPartialUpdate({
        key: this.lastNodeKey,
        value: field.getPartial(),
        delta: stash,
        field,
      });
    }
    await this.emitFinalUpdate(this.lastNodeKey, field);

    const currentNode = this.nodes[this.lastNodeKey];
    if (!currentNode.isEnd) {
      this.throwWithContext(
        `Node '${this.lastNodeKey}' is not an end node.`,
        LinePrefixParserError.Reason.NotEndNode,
      );
    }

    await Promise.allSettled(Object.values(this.nodes).map(({ field }) => field.end()));
    return this.finalState;
  }

  protected async emitPartialUpdate(data: InferCallbackValue<Callbacks<T>["partialUpdate"]>) {
    if (data.key in this.finalState) {
      this.throwWithContext(
        `Cannot update partial event for completed key '${data.key}'`,
        LinePrefixParserError.Reason.AlreadyCompleted,
      );
    }
    if (!(data.key in this.partialState)) {
      this.partialState[data.key] = "";
    }
    this.partialState[data.key] += data.delta;
    if (!this.options.silentNodes?.includes(data.key)) {
      await this.emitter.emit("partialUpdate", data);
    }
  }

  protected async emitFinalUpdate(key: StringKey<T>, field: ParserField<any, any>) {
    if (key in this.finalState) {
      this.throwWithContext(
        `Duplicated key '${key}'`,
        LinePrefixParserError.Reason.AlreadyCompleted,
      );
    }

    try {
      const value = field.get();
      this.finalState[key] = value;
      if (!this.options.silentNodes?.includes(key)) {
        await this.emitter.emit("update", {
          key,
          field,
          value,
        });
      }
    } catch (e) {
      if (e instanceof ZodError) {
        this.throwWithContext(
          `Value for '${key}' cannot be retrieved because it's value does not adhere to the appropriate schema.`,
          LinePrefixParserError.Reason.InvalidSchema,
          { errors: [e] },
        );
      }
      throw e;
    }
  }

  @Cache()
  protected get normalizedNodes() {
    return entries(this.nodes)
      .sort(([_, a], [__, b]) => a.prefix.length - b.prefix.length)
      .map(([key, value]) => {
        return [key, { lowerCasePrefix: value.prefix.toLowerCase(), ref: value }] as const;
      });
  }

  protected extractLine(line: string): ExtractedLine<T> | null {
    const trimmedLine = trimLeftSpaces(line);
    if (!trimmedLine) {
      return null;
    }

    for (const [key, node] of this.normalizedNodes) {
      const partial = node.lowerCasePrefix.length > trimmedLine.length;
      const [a, b] = partial
        ? [node.lowerCasePrefix, trimmedLine]
        : [trimmedLine, node.lowerCasePrefix];

      if (a.toLowerCase().startsWith(b.toLowerCase())) {
        return {
          key: key as StringKey<T>,
          value: partial ? trimmedLine : trimLeftSpaces(a.substring(b.length)),
          partial,
        };
      }
    }
    return null;
  }

  createSnapshot() {
    return {
      nodes: shallowCopy(this.nodes),
      lines: shallowCopy(this.lines),
      emitter: this.emitter,
      done: this.done,
      lastNodeKey: this.lastNodeKey,
      options: this.options,
    };
  }

  loadSnapshot(snapshot: ReturnType<typeof this.createSnapshot>) {
    Object.assign(this, snapshot);
  }
}

// eslint-disable-next-line @typescript-eslint/no-namespace
export namespace LinePrefixParser {
  export type infer<T extends Record<string, ParserNode<StringKey<T>, ParserField<any, any>>>> = {
    [K in keyof T]: ParserField.inferValue<T[K]["field"]>;
  };
  export type inferPartial<
    T extends Record<string, ParserNode<StringKey<T>, ParserField<any, any>>>,
  > = {
    [K in StringKey<T>]: ParserField.inferPartialValue<T[K]["field"]>;
  };

  export type inferCallback<T> =
    T extends LinePrefixParser<any> ? (T["emitter"] extends Emitter<infer L> ? L : never) : never;

  export type inferOutput<T> =
    T extends LinePrefixParser<infer P> ? LinePrefixParser.infer<P> : never;

  export type inferPartialOutput<T> =
    T extends LinePrefixParser<infer P> ? LinePrefixParser.inferPartial<P> : never;

  export type define<T extends Record<string, ParserField<any, any>>> = {
    [K in StringKey<T>]: ParserNode<StringKey<T>, T[K]>;
  };
}

File: python/beeai_framework/parsers/line_prefix.py
# Copyright 2025 IBM Corp.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from collections.abc import Callable
from enum import StrEnum
from typing import Any, NoReturn

from pydantic import BaseModel, InstanceOf, ValidationError

from beeai_framework.emitter import Emitter, EmitterInput
from beeai_framework.errors import FrameworkError
from beeai_framework.parsers.field import ParserField
from beeai_framework.utils.strings import trim_left_spaces

NEW_LINE_CHARACTER = "\n"


class LinePrefixParserUpdate(BaseModel):
    key: str
    value: Any
    delta: str
    field: InstanceOf[ParserField]


class LinePrefixParserLine(BaseModel):
    value: str
    new_line: bool


class LinePrefixParserExtractedLine(BaseModel):
    key: str
    value: str
    partial: bool


class LinePrefixParserOptions(BaseModel):
    fallback: Callable[[str], list[dict[str, str]]] | None = None
    end_on_repeat: bool = False
    wait_for_start_node: bool = False
    silent_nodes: list[str] | None = []


class LinePrefixParserNode(BaseModel):
    prefix: str
    next: list[str] = []
    field: InstanceOf[ParserField]
    is_start: bool = False
    is_end: bool = False


Nodes = dict[str, LinePrefixParserNode]
Customizer = Callable[[Nodes, LinePrefixParserOptions], tuple[Nodes, LinePrefixParserOptions]]


def _lines_to_string(lines: list["LinePrefixParserLine"]) -> str:
    result = ""
    for line in lines:
        if line.new_line:
            result += "\n" + line.value
        else:
            result += line.value
    return result


class LinePrefixParser:
    def __init__(self, nodes: Nodes, options: LinePrefixParserOptions | None = None) -> None:
        if options is None:
            options = LinePrefixParserOptions()
        self.nodes: Nodes = nodes
        self.options: LinePrefixParserOptions = options
        self.emitter: Emitter = Emitter(EmitterInput(creator=self, namespace=["parser", "line"]))
        self.lines: list[LinePrefixParserLine] = []
        self.excluded_lines: list[LinePrefixParserLine] = []
        self.done: bool = False
        self.last_node_key: str | None = None

        # final_state will contain the final parsed values.
        self.final_state: dict[str, Any] = {}
        # partial_state holds incremental updates.
        self.partial_state: dict[str, str] = {}

        has_start_node = False
        has_end_node = False
        for key, node in self.nodes.items():
            has_start_node = has_start_node or node.is_start
            has_end_node = has_end_node or node.is_end
            for next_key in node.next:
                if key == next_key:
                    raise ValueError(f"Node '{key}' cannot point to itself.")
                if next_key not in self.nodes:
                    raise ValueError(f"Node '{key}' contains a transition to non-existing node '{next_key}'.")
        if not has_start_node:
            raise ValueError("At least one start node must be provided!")
        if not has_end_node:
            raise ValueError("At least one end node must be provided!")

    def fork(self, customizer: Customizer) -> "LinePrefixParser":
        new_nodes, new_options = customizer(self.nodes, self.options)
        return LinePrefixParser(new_nodes, new_options)

    async def add(self, chunk: str) -> None:
        if not chunk or self.done:
            return

        parts = chunk.split(NEW_LINE_CHARACTER)
        for i, value in enumerate(parts):
            if i == 0:
                if not self.lines:
                    self.lines.append(LinePrefixParserLine(value=value, new_line=False))
                else:
                    self.lines[-1].value += value
            else:
                self.lines.append(LinePrefixParserLine(value=value, new_line=(len(parts) > 1)))

        while self.lines:
            line = self.lines[0]
            is_last_line = len(self.lines) == 1
            last_node = self.nodes[self.last_node_key] if self.last_node_key else None
            is_termination_node = last_node and last_node.is_end and (len(last_node.next) == 0)
            parsed_line: LinePrefixParserExtractedLine | None = None
            if not (is_termination_node or (last_node and not line.new_line)):
                parsed_line = self._extract_line(line.value)

            if is_last_line and ((parsed_line is not None and parsed_line.partial) or not line.value):
                break

            self.lines.pop(0)

            if parsed_line and not parsed_line.partial:
                assert parsed_line is not None
                if last_node:
                    if parsed_line.key not in last_node.next:
                        if parsed_line.key in self.final_state and self.options.end_on_repeat and last_node.is_end:
                            await self.end()
                            return
                        self.throw_with_context(
                            f"Transition from '{self.last_node_key}' to '{parsed_line.key}' does not exist!",
                            LinePrefixParserError.Reason.InvalidTransition,
                            extra={"line": line},
                        )

                    assert self.last_node_key is not None
                    await self._emit_final_update(self.last_node_key, last_node.field)
                elif not self.nodes[parsed_line.key].is_start:
                    if not self.options.wait_for_start_node:
                        self.throw_with_context(
                            f'Parsed text line corresponds to a node "{parsed_line.key}" which is not a start node!',
                            LinePrefixParserError.Reason.NotStartNode,
                            extra={"line": line},
                        )
                    self.excluded_lines.append(line)
                    continue

                node = self.nodes[parsed_line.key]
                node.field.write(parsed_line.value)
                await self._emit_partial_update(
                    LinePrefixParserUpdate(
                        key=parsed_line.key, value=parsed_line.value, delta=parsed_line.value, field=node.field
                    )
                )
                self.last_node_key = parsed_line.key
            elif self.last_node_key:
                if not self.nodes[self.last_node_key].field.raw:
                    line.value = trim_left_spaces(line.value)
                if line.new_line:
                    line.value = NEW_LINE_CHARACTER + line.value
                node = self.nodes[self.last_node_key]
                node.field.write(line.value)
                await self._emit_partial_update(
                    LinePrefixParserUpdate(
                        key=self.last_node_key, value=node.field.get_partial(), delta=line.value, field=node.field
                    )
                )
            else:
                self.excluded_lines.append(line)

    def throw_with_context(self, message: str, reason: str, extra: dict[str, Any] | None = None) -> NoReturn:
        extra = extra or {}
        context_lines = self.lines + ([extra["line"]] if "line" in extra else [])
        context = {
            "lines": _lines_to_string(context_lines),
            "excludedLines": _lines_to_string(self.excluded_lines),
            "finalState": self.final_state,
            "partialState": self.partial_state,
        }
        full_message = f"The generated output does not adhere to the schema.\n{message}"
        raise LinePrefixParserError(full_message, reason=reason, context=context)

    async def end(self) -> dict[str, Any]:
        if self.done:
            return self.final_state

        if not self.last_node_key and self.options.fallback:
            stash = _lines_to_string(self.excluded_lines + self.lines)
            self.excluded_lines.clear()
            self.lines.clear()
            fallback_nodes = self.options.fallback(stash)
            fallback_text = NEW_LINE_CHARACTER.join(
                [self.nodes[node["key"]].prefix + node["value"] for node in fallback_nodes]
            )
            await self.add(fallback_text)

        self.done = True

        if not self.last_node_key:
            self.throw_with_context("Nothing valid has been parsed yet!", LinePrefixParserError.Reason.NoDataReceived)

        stash = _lines_to_string(self.lines)
        self.lines.clear()
        field = self.nodes[self.last_node_key].field
        if stash:
            field.write(stash)
            await self._emit_partial_update(
                LinePrefixParserUpdate(key=self.last_node_key, value=field.get_partial(), delta=stash, field=field)
            )
        await self._emit_final_update(self.last_node_key, field)
        current_node = self.nodes[self.last_node_key]
        if not current_node.is_end:
            self.throw_with_context(
                f"Node '{self.last_node_key}' is not an end node.",
                LinePrefixParserError.Reason.NotEndNode,
            )

        for node in self.nodes.values():
            node.field.end()

        return self.final_state

    async def _emit_partial_update(self, data: LinePrefixParserUpdate) -> None:
        if data.key in self.final_state:
            self.throw_with_context(
                f"Cannot update partial event for completed key '{data.key}'",
                LinePrefixParserError.Reason.AlreadyCompleted,
            )
        if data.key not in self.partial_state:
            self.partial_state[data.key] = ""
        self.partial_state[data.key] += data.delta
        if not (self.options.silent_nodes and data.key in self.options.silent_nodes):
            await self.emitter.emit("partial_update", data)

    async def _emit_final_update(self, key: str, field: ParserField) -> None:
        if key in self.final_state:
            self.throw_with_context(f"Duplicated key '{key}'", LinePrefixParserError.Reason.AlreadyCompleted)
        try:
            value = field.get()
            self.final_state[key] = value.model_dump()
            if not (self.options.silent_nodes and key in self.options.silent_nodes):
                await self.emitter.emit("update", LinePrefixParserUpdate(key=key, value=value, field=field, delta=""))
        except ValidationError:
            self.throw_with_context(
                f"Value for '{key}' cannot be retrieved because its value does not adhere to the appropriate schema.",
                LinePrefixParserError.Reason.InvalidSchema,
            )

    @property
    def _normalized_nodes(self) -> list[tuple[str, dict[str, Any]]]:
        sorted_nodes = sorted(self.nodes.items(), key=lambda item: len(item[1].prefix))
        return [(key, {"lowerCasePrefix": node.prefix.lower(), "ref": node}) for key, node in sorted_nodes]

    def _extract_line(self, line: str) -> LinePrefixParserExtractedLine | None:
        trimmed_line = trim_left_spaces(line)
        if not trimmed_line:
            return None

        for key, node_info in self._normalized_nodes:
            lower_case_prefix = node_info["lowerCasePrefix"]
            partial = len(lower_case_prefix) > len(trimmed_line)
            if partial:
                a, b = lower_case_prefix, trimmed_line
            else:
                a, b = trimmed_line, lower_case_prefix

            if a.lower().startswith(b.lower()):
                value = trimmed_line if partial else trim_left_spaces(trimmed_line[len(b) :])
                return LinePrefixParserExtractedLine(key=key, value=value, partial=partial)

        return None


class LinePrefixParserError(FrameworkError):
    class Reason(StrEnum):
        NoDataReceived = "NoDataReceived"
        InvalidTransition = "InvalidTransition"
        NotStartNode = "NotStartNode"
        NotEndNode = "NotEndNode"
        AlreadyCompleted = "AlreadyCompleted"
        InvalidSchema = "InvalidSchema"

    def __init__(self, message: str, *, context: dict[str, Any], reason: str) -> None:
        super().__init__(message, is_fatal=True, is_retryable=True)
        self.context = context
        self.reason = reason
